### PR TITLE
UC-179 Make sure we can retirect back to the Special:Preferences page

### PR DESCRIPTION
**Pull request by garthwebb**
Found wfGetReturnToParam that was supposed to be used on this type of redirect.  It seems
like Marooned added this and updated a bunch of usages but forgot this one.

Also had to make sure UserLoginFacebook.js responded to the returnto parameter which it wasn't doing previously.

https://wikia-inc.atlassian.net/browse/UC-179

### CODE REVIEW

File: extensions/wikia/UserLogin/js/UserLoginFacebook.js [hunk 1]
```diff
@@ -170,19 +170,28 @@
...
+					returnTo = returnUrl.getVal('returnto');
 
+				// See if we were passed a returnto URL.  If not and we're on a special page we shouldn't
+				// stay on, set returnto to be the main page
 				if (returnTo) {
-					qString.setPath(wgArticlePath.replace('$1', returnTo));
+					returnUrl.removeVal('returnto');
+				} else if (
+					wgCanonicalSpecialPageName &&
+					wgCanonicalSpecialPageName.match(/Userlogin|Userlogout|UserSignup/)
```
Comment (lizlux) [inline extensions/wikia/UserLogin/js/UserLoginFacebook.js]: Per our discussion, let's use some version of the existing back end logic for this.

File: extensions/wikia/FacebookClient/FacebookClient.class.php [hunk 2]
```diff
@@ -369,47 +369,6 @@ public function logout() {
 	}
 
 	/**
-	 * Check if we should redirect back to the specified page by comparing it to this black list
```
Comment (garthwebb) [inline extensions/wikia/FacebookClient/FacebookClient.class.php]: This code was duplicate functionality.  Consolidated into the UserLoginHelper.class.php method, `getRedirectUrl()`

File: extensions/wikia/FacebookClient/tests/FacebookClientTest.php [hunk 3]
```diff
@@ -1,47 +0,0 @@
-<?
```
Comment (garthwebb) [inline extensions/wikia/FacebookClient/tests/FacebookClientTest.php]: Moved all these tests into `UserLoginTest.php` and updated the method they call.

File: extensions/wikia/UserLogin/FacebookSignupController.class.php [hunk 4]
```diff
@@ -33,35 +37,58 @@ public function index() {
 		$user = FacebookClient::getInstance()->getWikiaUser( $fbUserId );
 
 		if ( ( $user instanceof User ) && ( $fbUserId !== 0 ) ) {
-			$this->errorMsg = '';
+			$errorMsg = '';
```
Comment (garthwebb) [inline extensions/wikia/UserLogin/FacebookSignupController.class.php]: Setting all the response data in one go in each conditional, so can't set a single val here.

File: extensions/wikia/UserLogin/FacebookSignupController.class.php [hunk 5]
```diff
@@ -279,27 +301,25 @@ protected function getValidFbUserId() {
 	 * @return null|User
 	 */
 	protected function getValidWikiaUser( $wikiaUserName, $wikiaPassword ) {
```
Comment (garthwebb) [inline extensions/wikia/UserLogin/FacebookSignupController.class.php]: This nested conditional was more confusing than helpful.  Split it out into individual checks.

File: extensions/wikia/UserLogin/UserLoginHelper.class.php [hunk 6]
```diff
@@ -4,22 +4,19 @@
  * User Login Helper
  * @author Hyun
  * @author Saipetch
- *
  */
 class UserLoginHelper extends WikiaModel {
 
-	protected static $instance = NULL;
```
Comment (garthwebb) [inline extensions/wikia/UserLogin/UserLoginHelper.class.php]: This didn't seem to be used anywhere.

File: extensions/wikia/UserLogin/UserLoginHelper.class.php [hunk 7]
```diff
@@ -4,22 +4,19 @@
...
-
 	const LIMIT_EMAIL_CHANGES = 5;
 	const LIMIT_EMAILS_SENT = 5;
 	const LIMIT_AVATARS = 4;
 	const LIMIT_WIKIS = 3;
 
-	const WIKIA_CITYID_COMMUNITY = 177;
```
Comment (garthwebb) [inline extensions/wikia/UserLogin/UserLoginHelper.class.php]: Don't redefine this; its already in WikiFactory

File: extensions/wikia/UserLogin/UserLoginHelper.class.php [hunk 8]
```diff
@@ -133,97 +133,82 @@ protected function addUserToUserList( $userId, &$userList ) {
...
 
 	/**
-	 * get a list of random popular wikis
-	 * @param integer $require (number of wikis)
-	 * @return array $wikis
+	 * Redirect the user to the appropriate page after login.
+	 *
+	 * @requestParam string returnUrl
 	 */
-	public function getRandomWikis( $require=self::LIMIT_WIKIS ) {
```
Comment (garthwebb) [inline extensions/wikia/UserLogin/UserLoginHelper.class.php]: Got rid of this unused function after checking with Cristina on the functionality it was providing.  I guess Tor added this early return in 2012 while investigating a slow query here and never finished up.  Blame Tor!  I cleaned up everything above that was calling this.

File: extensions/wikia/UserLogin/UserSignupSpecialController.class.php [hunk 9]
```diff
@@ -100,7 +100,6 @@ public function signupForm () {
 
 		//fb#38260 -- removed uselang
 		$this->avatars = $this->userLoginHelper->getRandomAvatars();
-		$this->popularWikis = $this->userLoginHelper->getRandomWikis();
```
Comment (garthwebb) [inline extensions/wikia/UserLogin/UserSignupSpecialController.class.php]: Cleanup from the Blame Tor! annotation further up.

File: extensions/wikia/UserLogin/js/UserLoginFacebook.js [hunk 10]
```diff
@@ -163,30 +167,6 @@
 		},
 
 		/**
-		 * This runs after has signed in with facebook and is already registered with Wikia.
```
Comment (garthwebb) [inline extensions/wikia/UserLogin/js/UserLoginFacebook.js]: We're doing all this in the backend now and just passing the exact URL to redirect to back with the AJAX response.

File: extensions/wikia/UserLogin/templates/UserSignupSpecial_signupForm.php [hunk 11]
```diff
@@ -157,11 +156,6 @@
 				<p><?= wfMessage( 'usersignup-marketing-community' )->escaped() ?></p>
 			</div>
 			<div class="benefit">
-				<ul class="wikis">
-					<? foreach( $popularWikis as $wiki ) { ?>
-					<li class="wiki"><img src="<?= $wiki ?>" width="65" height="18"></li>
-					<? } ?>
-				</ul>
```
Comment (garthwebb) [inline extensions/wikia/UserLogin/templates/UserSignupSpecial_signupForm.php]: This is more Blame Tor! cleanup.

File: extensions/wikia/UserLogin/tests/UserLoginTest.php [hunk 12]
```diff
@@ -1,782 +1,869 @@
 <?php
```
Comment (garthwebb) [inline extensions/wikia/UserLogin/tests/UserLoginTest.php]: Ok.  This entire file was tabbed in one stop for some reason.  I fixed the tabbing but it makes this messy.  I'll note the only addition I made to this file below.

File: extensions/wikia/UserLogin/tests/UserLoginTest.php [hunk 13]
```diff
@@ -1,782 +1,869 @@
...
-			];
-		}
+				'wfArrayToCGI' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
+				'wfCgiToArrayResult' => [],
+				'expected' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
+				'message' => 'Query with returntoquery',
+			],
+		];
+	}
+
```
Comment (garthwebb) [inline extensions/wikia/UserLogin/tests/UserLoginTest.php]: Ok, finally!  This method and its data provider are the two things I added.

Comment (armonr) [inline extensions/wikia/UserLogin/tests/UserLoginTest.php]: Since `getReturnToUrl()` is removed, `testGetRedirectURL` and `getRedirectURLProvider` would be better names.

File: extensions/wikia/UserLogin/FacebookSignupController.class.php [hunk 14]
```diff
@@ -13,10 +13,14 @@ class FacebookSignupController extends WikiaController {
...
 	protected $fbClientFactory;
 
+	/** @var \UserLoginHelper $userLoginHelper */
+	protected $userLoginHelper;
+
 	public function __construct() {
 		parent::__construct();
 
 		$this->fbClientFactory = new \FacebookClientFactory();
+		$this->userLoginHelper = new \UserLoginHelper();
```
Comment (armonr) [inline extensions/wikia/UserLogin/FacebookSignupController.class.php]: :sunny:

File: extensions/wikia/UserLogin/FacebookSignupController.class.php [hunk 15]
```diff
@@ -279,27 +301,25 @@ protected function getValidFbUserId() {
...
+			return null;
+		}
+
+		if ( !$wikiaPassword ) {
+			$this->setErrorResponse( 'userlogin-error-wrongpasswordempty', [], self::SIGNUP_PASSWORD_KEY );
+			return null;
+		}
+
+		$user = \User::newFromName( $wikiaUserName );
+		if ( empty( $user ) ) {
```
Comment (armonr) [inline extensions/wikia/UserLogin/FacebookSignupController.class.php]: Any reason for using `empty()` here?

File: extensions/wikia/UserLogin/UserLoginHelper.class.php [hunk 16]
```diff
@@ -133,97 +133,82 @@ protected function addUserToUserList( $userId, &$userList ) {
...
-				if ( count($wikis) >= $require )
-					break;
-			}
-
-			$this->wg->Memc->set( $memKey, $wikis, 60*60*24 );
-		}
-
-		return $wikis;
- **/
+	public function doRedirect() {
```
Comment (armonr) [inline extensions/wikia/UserLogin/UserLoginHelper.class.php]: Does this need to also carry optional parameters `$extraReturnToQuery = '', $cbVal = 0 ` to pass on to `getRedirectUrl`?

File: extensions/wikia/UserLogin/UserLoginHelper.class.php [hunk 17]
```diff
@@ -133,97 +133,82 @@ protected function addUserToUserList( $userId, &$userList ) {
...
-		if ( empty($popularWikis) ) {
-			$popularWikis = array_keys( DataMartService::getTopWikisByPageviews( DataMartService::PERIOD_ID_MONTHLY, $limit ) );
+		if ( $this->isInvalidLoginRedirect( $titleObj ) ) {
+			$titleObj = Title::newMainPage();
+		}
 
-			if ( empty($popularWikis) ) {
-				$popularWikis[] = self::WIKIA_CITYID_COMMUNITY;
-			}
+		$cbParam = "cb=" . ( $cbVal ? $cbVal : rand( 1, 10000 ) );
```
Comment (armonr) [inline extensions/wikia/UserLogin/UserLoginHelper.class.php]: single quotes :)

File: extensions/wikia/UserLogin/UserLoginHelper.class.php [hunk 18]
```diff
@@ -401,10 +396,11 @@ public function incrMemc( $memKey, $period=86400 ) {
 
 	/**
 	 * clear password-throttle cache
+	 *
 	 * @param string $username
 	 */
 	public function clearPasswordThrottle( $username ) {
-		$key = wfMemcKey( 'password-throttle', wfGetIP(), md5($username) );
+		$key = wfMemcKey( 'password-throttle', $this->wg->Request->getIP(), md5( $username ) );
```
Comment (armonr) [inline extensions/wikia/UserLogin/UserLoginHelper.class.php]: This seems like a good candidate for a standalone function.

File: extensions/wikia/UserLogin/FacebookSignupController.class.php (hunk 15 above)
Comment (garthwebb) [inline extensions/wikia/UserLogin/FacebookSignupController.class.php]: Nothing more than its sort of an unwritten standard at Wikia to use `empty()` to check for unset rather than applying logic like this directly, at least with a single variable.  Also, I like to test for positives where possible (empty is positive in the way that it asserts something rather than asserting not-something).

File: extensions/wikia/UserLogin/UserLoginHelper.class.php (hunk 16 above)
Comment (garthwebb) [inline extensions/wikia/UserLogin/UserLoginHelper.class.php]: It doesn't as the only use case for this is for `getRedirectUrl` but I can add it for completeness.

File: extensions/wikia/UserLogin/UserLoginHelper.class.php (hunk 18 above)
Comment (garthwebb) [inline extensions/wikia/UserLogin/UserLoginHelper.class.php]: You're talking about `getIP`?  Isn't this a stand-alone function already?

File: extensions/wikia/UserLogin/tests/UserLoginTest.php (hunk 13 above)
Comment (garthwebb) [inline extensions/wikia/UserLogin/tests/UserLoginTest.php]: Yup, that makes sense.

File: extensions/wikia/UserLogin/UserLoginHelper.class.php (hunk 18 above)
Comment (armonr) [inline extensions/wikia/UserLogin/UserLoginHelper.class.php]: I was thinking the whole line because it is repeated.
```php
protected function getPasswordThrottleCacheKey( $username ) {
	return wfMemcKey( 'password-throttle', $this->wg->Request->getIP(), md5( $username ) );
}
```

File: extensions/wikia/UserLogin/FacebookSignupController.class.php (hunk 15 above)
Comment (armonr) [inline extensions/wikia/UserLogin/FacebookSignupController.class.php]: Sure. Although for high level variables, as opposed to e.g `$foo['bar']`, `!` operator is slightly more efficient and cleaner to read :)

File: extensions/wikia/UserLogin/tests/UserLoginTest.php [hunk 19]
```diff
@@ -1,782 +1,869 @@
...
+				'foo=bar',
+				'fbconnected=1',
+				456,
+				Title::newFromText( 'Foo' ),
+				'foo=bar&fbconnected=1&cb=456'
+			],
+
+			// test forbidden return-to page w/o query string
+			[
+				'Special:Signup',
```
Comment (armonr) [inline extensions/wikia/UserLogin/tests/UserLoginTest.php]: 'Special:Signup' redirects to 'Special:UserLogin'. Should it be changed everywhere or is this a NS defined somewhere?
Also is there a case for `:UserSignup`?

**Comment by armonr** [comment]
Please retarget to uconn-5 branch

File: extensions/wikia/UserLogin/tests/UserLoginTest.php (hunk 19 above)
Comment (garthwebb) [inline extensions/wikia/UserLogin/tests/UserLoginTest.php]: This is not testing where the user is sent on any particular page in these tests, only that if the user happens to have a forbidden page as their `returnto` that they are sent to the main page.

**Comment by lizlux** [comment]
Using https://github.com/Wikia/app/pull/6024/files instead.
